toolbox/relnotes/main.go

```diff
@@ -0,0 +1,757 @@
+	}
+}
+
+// minorReleases performs a minor (vX.Y.0) release by fetching the release template and aggregate
+// previous release in series.
+func minorRelease(f *os.File, release, draftURL, changelogURL string) {
+
+	// Check for draft and use it if available
+	log.Printf("Checking if draft release notes exist for %s...", release)
+
+	resp, err := http.Get(draftURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	// TODO: find a better way to tell failed response
+	if err == nil && (resp.StatusCode == 200 || resp.StatusCode == 304) {
+		log.Printf("Draft found - using for release notes...")
+		_, err = io.Copy(f, resp.Body)
+		if err != nil {
+			log.Printf("error during copy to file: %s", err)
+			return
+		}
+		f.WriteString("\n")
+	} else {
+		log.Printf("No draft found - creating generic template...")
+		f.WriteString("## Major Themes\n\n* TBD\n\n## Other notable improvements\n\n* TBD\n\n## Known Issues\n\n* TBD\n\n## Provider-specific Notes\n\n* TBD\n\n")
+	}
+
+	// Aggregate all previous release in series
+	f.WriteString(fmt.Sprintf("### Previous Release Included in %s\n\n", release))
+
+	// Regexp Example:
+	// Assume the release tag is v1.7.0, this regexp matches "- [v1.7.0-" in
+	//     "- [v1.7.0-rc.1](#v170-rc1)"
+	//     "- [v1.7.0-beta.2](#v170-beta2)"
+	reAnchor, _ := regexp.Compile(fmt.Sprintf("- \\[%s-", release))
+
+	resp, err = http.Get(changelogURL)
+	if err == nil {
+		defer resp.Body.Close()
+	}
+
+	// TODO: find a better way to tell failed response
+	if err == nil && (resp.StatusCode == 200 || resp.StatusCode == 304) {
+		buf := new(bytes.Buffer)
+		buf.ReadFrom(resp.Body)
+		for _, line := range strings.Split(buf.String(), "\n") {
+			if anchor := reAnchor.FindStringSubmatch(line); anchor != nil {
+				f.WriteString(line + "\n")
+			}
+		}
+		f.WriteString("\n")
+	}
+
+}
+
+// patchRelease performs a patch (vX.Y.Z) release by printing out all the related changes.
+func patchRelease(f *os.File, info *ReleaseInfo) {
+	// Release note for different labels (TODO: "release-note" label for now since "experimental" and
+	// "action" are deprecated)
+	f.WriteString(fmt.Sprintf("## Changelog since %s\n\n", info.startTag))
+
+	if len(info.releaseActionRequiredPRs) > 0 {
+		f.WriteString("### Action Required\n\n")
+		for _, pr := range info.releaseActionRequiredPRs {
+			f.WriteString(fmt.Sprintf("* %s (#%d, @%s)\n", extractReleaseNote(info.prMap[pr]), pr, *info.prMap[pr].User.Login))
+		}
+		f.WriteString("\n")
+	}
+
+	if len(info.releasePRs) > 0 {
+		f.WriteString("### Other notable changes\n\n")
+		for _, pr := range info.releasePRs {
+			f.WriteString(fmt.Sprintf("* %s (#%d, @%s)\n", extractReleaseNote(info.prMap[pr]), pr, *info.prMap[pr].User.Login))
+		}
+		f.WriteString("\n")
+	} else {
+		f.WriteString("**No notable changes for this release**\n\n")
+	}
+}
+
+// extractReleaseNote tries to fetch release note from PR body, otherwise uses PR title.
+func extractReleaseNote(pr *github.Issue) string {
```

Comment:
change to name to "extractReleaseNoteFromPR"

Reply:
Done